### PR TITLE
Rust Refactor Stage 4: Rewrite Rust graph runtime to use new APIs

### PR DESCRIPTION
**Pull request by jroesch**
This is the fourth stage of the Rust rewrite which ports the graph-rt to the new API. The final stage after this will be turning on CI for the new bindings, updating docs, and deprecating the old bindings. This depends on #5764.

### CODE REVIEW

**Comment by jroesch** [comment]
@robo-corg @tqchen this is like a straight copy of the old `runtime` crate with slight modifications to use the new `tvm-sys`. After this patch lands I can send a final one to switch CI over to the new bindings and drop support for the old ones.

File: rust/tvm-graph-rt/src/array.rs [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+            Self {
+                data: storage,
+                ctx: Context::default(),
+                dtype,
+                size,
+                shape,
+                strides: if dlt.strides.is_null() {
+                    None
+                } else {
+                    Some(slice::from_raw_parts_mut(dlt.strides as *mut usize, size).to_vec())
```
Comment (leonwanghui) [inline rust/tvm-graph-rt/src/array.rs:365]:
```suggestion
                    Some(slice::from_raw_parts_mut(dlt.strides as *mut usize, shape.len()).to_vec())
```

**Comment by kazum** [comment]
@jroesch Can we enable derive_default for bindgen?  Otherwise, test_wasm32 fails with tvm-sys because of the generated padding field.

The following change looks necessary which existed in the previous implementation.

```diff
diff --git a/rust/tvm-graph-rt/src/array.rs b/rust/tvm-graph-rt/src/array.rs
index 38519bd..8209b59 100644
--- a/rust/tvm-graph-rt/src/array.rs
+++ b/rust/tvm-graph-rt/src/array.rs
@@ -288,6 +288,7 @@ impl<'a> Tensor<'a> {
                 self.strides.as_ref().unwrap().as_ptr()
             } as *mut i64,
             byte_offset: 0,
+            ..Default::default()
         }
     }
 }
diff --git a/rust/tvm-sys/build.rs b/rust/tvm-sys/build.rs
index 85e16be..01d2934 100644
--- a/rust/tvm-sys/build.rs
+++ b/rust/tvm-sys/build.rs
@@ -54,6 +54,7 @@ fn main() {
         .layout_tests(false)
         .derive_partialeq(true)
         .derive_eq(true)
+        .derive_default(true)
         .generate()
         .expect("unable to generate bindings")
         .write_to_file(PathBuf::from("src/c_runtime_api.rs"))
diff --git a/rust/tvm-sys/src/array.rs b/rust/tvm-sys/src/array.rs
index 1627e9e..5d09d86 100644
--- a/rust/tvm-sys/src/array.rs
+++ b/rust/tvm-sys/src/array.rs
@@ -48,6 +48,7 @@ macro_rules! impl_dltensor_from_ndarray {
                     shape: arr.shape().as_ptr() as *const i64 as *mut i64,
                     strides: arr.strides().as_ptr() as *const i64 as *mut i64,
                     byte_offset: 0,
+                    ..Default::default()
                 }
             }
         }
```

File: rust/tvm-sys/src/datatype.rs [hunk 2]
```diff
@@ -73,6 +73,18 @@ impl DataType {
...
+    pub const fn int(bits: u8, lanes: u16) -> DataType {
+        DataType::new(DL_INT_CODE, bits, lanes)
+    }
+
+    pub const fn float(bits: u8, lanes: u16) -> DataType {
+        DataType::new(DL_FLOAT_CODE, bits, lanes)
+    }
+
+    pub const fn uint(bits: u8, lanes: u16) -> DataType {
+        DataType::new(DL_FLOAT_CODE, bits, lanes)
```
Comment (robo-corg) [inline rust/tvm-sys/src/datatype.rs:86]: Should this be `DL_FLOAT_CODE`?
```suggestion
        DataType::new(DL_UINT_CODE, bits, lanes)
```

File: rust/tvm-graph-rt/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+repository = "https://github.com/apache/incubator-tvm"
+readme = "README.md"
+keywords = ["tvm"]
+categories = ["api-bindings", "science"]
+authors = ["TVM Contributors"]
+edition = "2018"
+
+[dependencies]
+crossbeam = "0.7.3"
+failure = "0.1"
```
Comment (binarybana) [inline rust/tvm-graph-rt/Cargo.toml:32]: Not `anyhow` or `thiserror`?

File: rust/tvm-graph-rt/tests/test_tvm_basic/src/build_test_lib.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+from tvm import te
+
+def main():
+    n = te.var('n')
+    A = te.placeholder((n,), name='A')
+    B = te.placeholder((n,), name='B')
+    C = te.compute(A.shape, lambda *i: A(*i) + B(*i), name='C')
+    s = tvm.te.create_schedule(C.op)
+    s[C].parallel(s[C].op.axis[0])
+    print(tvm.lower(s, [A, B, C], simple_mode=True))
```
Comment (binarybana) [inline rust/tvm-graph-rt/tests/test_tvm_basic/src/build_test_lib.py:34]: Do you want this `print` statement here?

File: rust/tvm-graph-rt/tests/test_wasm32/src/build_test_lib.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+from tvm import te
+
+def main():
+    n = te.var('n')
+    A = te.placeholder((n,), name='A')
+    B = te.placeholder((n,), name='B')
+    C = te.compute(A.shape, lambda *i: A(*i) + B(*i), name='C')
+    s = tvm.te.create_schedule(C.op)
+    s[C].parallel(s[C].op.axis[0])
+    print(tvm.lower(s, [A, B, C], simple_mode=True))
```
Comment (binarybana) [inline rust/tvm-graph-rt/tests/test_wasm32/src/build_test_lib.py:34]: Another print.

**Comment by jroesch** [comment]
> @jroesch Can we enable derive_default for bindgen? Otherwise, test_wasm32 fails with tvm-sys because of the generated padding field.
> 
> The following change looks necessary which existed in the previous implementation.
> 
> ```diff
> diff --git a/rust/tvm-graph-rt/src/array.rs b/rust/tvm-graph-rt/src/array.rs
> index 38519bd..8209b59 100644
> --- a/rust/tvm-graph-rt/src/array.rs
> +++ b/rust/tvm-graph-rt/src/array.rs
> @@ -288,6 +288,7 @@ impl<'a> Tensor<'a> {
>                  self.strides.as_ref().unwrap().as_ptr()
>              } as *mut i64,
>              byte_offset: 0,
> +            ..Default::default()
>          }
>      }
>  }
> diff --git a/rust/tvm-sys/build.rs b/rust/tvm-sys/build.rs
> index 85e16be..01d2934 100644
> --- a/rust/tvm-sys/build.rs
> +++ b/rust/tvm-sys/build.rs
> @@ -54,6 +54,7 @@ fn main() {
>          .layout_tests(false)
>          .derive_partialeq(true)
>          .derive_eq(true)
> +        .derive_default(true)
>          .generate()
>          .expect("unable to generate bindings")
>          .write_to_file(PathBuf::from("src/c_runtime_api.rs"))
> diff --git a/rust/tvm-sys/src/array.rs b/rust/tvm-sys/src/array.rs
> index 1627e9e..5d09d86 100644
> --- a/rust/tvm-sys/src/array.rs
> +++ b/rust/tvm-sys/src/array.rs
> @@ -48,6 +48,7 @@ macro_rules! impl_dltensor_from_ndarray {
>                      shape: arr.shape().as_ptr() as *const i64 as *mut i64,
>                      strides: arr.strides().as_ptr() as *const i64 as *mut i64,
>                      byte_offset: 0,
> +                    ..Default::default()
>                  }
>              }
>          }
> ```

Yeah I will turn it back on, there are just so many changes to juggle I was bound to make a mistake or two.

**Comment by jroesch** [comment]
@binarybana and @robo-corg many of the things you guys identified existed in the previous graph runtime, it would be my preference to land this port which just makes it compile against the new bindings, and then bring other patches to improve the code in follow up PRs. thoughts?

**Comment by binarybana** [comment]
Yes, that works for me, which is why I approved the PR despite my comments.

File: rust/tvm-graph-rt/src/array.rs (hunk 1 above)
Comment (jroesch) [reply rust/tvm-graph-rt/src/array.rs:366]: I don't think this is correct, as the field currently is passing the number of elements not the rank of the tensor.

**Comment by robo-corg** [comment]
> existed in the previous graph runtime, it would be my preference to land this port which just makes it compile against the new bindings, and then bring other patches to improve the code in follow up PRs. thoughts?

That works great!

File: rust/tvm-graph-rt/src/array.rs (hunk 1 above)
Comment (leonwanghui) [reply rust/tvm-graph-rt/src/array.rs:366]: From this figure (https://i.stack.imgur.com/oQQVI.png), I think the length of `strides` should be equal to the rank of the tensor (ie. dimension).

**Pull request merged**